File: util/python/src/aiddl_util/request/request_caller.py
class CallRequestFunction:
    def __init__(self, call, req_handler):
        self.C = call
        self.req_handler = req_handler

    def __call__(self, x):
        request = x[0]
        module = x[1]
        return_entry_name = x[2]

        self.C.add_module(module)
        self.req_handler.satisfy_request(request, module)

        r = self.C.get_entry(return_entry_name, module=module).get_value()
        return r

File: util/java/src/main/java/org/aiddl/util/java/request/RequestCaller.java
package org.aiddl.util.java.request;

import org.aiddl.core.java.container.Container;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Term;

/**
 * Function that calls a request. This allows to make AIDDL requests available as packaged functions.
 * 
 * @author Uwe Koeckemann
 *
 */
public class RequestCaller implements Function {
	
	RequestHandler rHandler;
	Container C;
	
	public RequestCaller ( Container C, RequestHandler rHandler ) {
		this.rHandler = rHandler;
		this.C = C;
	}

	@Override
	public Term apply(Term args) {
		Term request = args.get(0);
		Term module = args.get(1);
		Term returnEntryName = args.get(2);
		
		this.C.addModule(module);

		this.rHandler.satisfyRequest(request, this.C, module);
		
		Term r = this.C.getEntry(module, returnEntryName).getValue();
		
		return r;
	}
}
